mautic/mautic
# Prevent embedding the same image multiple times in an email when image embedding is enabled

To avoid embedding the same image multiple times, as described in https://github.com/mautic/mautic/issues/8995

| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | 3.0 
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | no
| Related user documentation PR URL      | 
| Related developer documentation PR URL | 
| Issue(s) addressed                     | Fixes #8995

This PR helps to avoid embedding the same image multiple times if it is referenced to more than once in the email.

When email is configured to get the images embedded, if the same image is referenced multiple times in the message, it gets added multiple times, even when only one reference will be used.

In function convertEmbedImages in https://github.com/mautic/mautic/blob/staging/app/bundles/EmailBundle/Helper/MailHelper.php we can see that for every match we get this executed:
$this->embedImagesReplaces[$match] = $this->message->embed(\Swift_Image::fromPath($match));

When the match is repeated, the key will be overwritten in $this->embedImagesReplaces, but message->embed will still be executed and the same image will be embedded multiple times.

I think before embedding the image you should check if key already exists.

## Steps to reproduce

1. Configure email to embed images in base64.
2. Create an email that references the same image several times.
3. Send the email and review the source code. You should see the image being embedded multiple times rather than embedded once and referenced to the same image.

## Steps to test the PR

1. Apply this PR
2. Configure email to embed images in base64.
3. Create an email that references the same image several times.
4. Send the email and review the source code. You should only see the image embedded once, not multiple times, and the image should still work in all the places it is referenced.

**Please ensure you also test that sending emails with multiple images and images being referenced in multiple places is still working with base64 disabled, to ensure that this PR does not have any untoward effects outside these conditions**
